spec/models/annotation_text_spec.rb

```diff
@@ -24,4 +24,135 @@
+  let(:text_without_deduction) { create(:annotation_text, annotation_category: category_without_criteria) }
+
+  describe 'validation of deduction' do
+    it 'is invalid with a deduction greater than the annotation_category\'s flexible_criterion\'s max mark' do
+      deductive_text.deduction = 4.0
+      expect(deductive_text).to_not be_valid
+    end
+
+    it 'is invalid with a deduction less than 0' do
+      deductive_text.deduction = -1.0
+      expect(deductive_text).to_not be_valid
+    end
+
+    it 'is valid with a numerical deduction if it does belong to flexible criteria through annotation category' do
+      expect(deductive_text).to be_valid
+    end
+
+    it 'is invalid with a nil deduction if it does belong to flexible criteria through annotation category' do
+      deductive_text.deduction = nil
+      expect(deductive_text).to_not be_valid
+    end
+
+    it 'is valid with a nil deduction if it does not belong to flexible criteria through annotation category' do
+      expect(text_without_deduction).to be_valid
+    end
+
+    it 'is invalid with a numerical deduction if it does not belong to flexible criteria through annotation category' do
+      text_without_deduction.deduction = 1.0
+      expect(text_without_deduction).to_not be_valid
+    end
+  end
+
+  describe 'callbacks' do
+    it 'prevent an update of deduction if any results are released' do
+      assignment.groupings.first.current_result.update!(released_to_students: true)
+      expect(deductive_text.update(deduction: 2.0)).to be(false)
+    end
+
+    it 'prevent an update of content if any results are released and there is a deduction' do
+      assignment.groupings.first.current_result.update!(released_to_students: true)
+      expect(deductive_text.update(content: 'Do not plagiarize!')).to be(false)
+    end
+
+    it 'do not prevent an update of content if any results are released and there is no deduction' do
+      expect(deductive_text.update(content: 'Do not plagiarize!')).to be(true)
+    end
+
+    it 'prevent a destruction of the annotation_text if any results are released and there is a deduction value' do
+      assignment.groupings.first.current_result.update!(released_to_students: true)
+      expect(deductive_text.destroy).to be(false)
+    end
+
+    it 'do not prevent a destruction of the annotation_text if no results are released even with a deduction value' do
+      expect(deductive_text.destroy).to eq(deductive_text)
+    end
+  end
+
+  describe '#update_mark_deductions' do
+    it 'updates the mark associated with its annotation category\'s flexible criterion' \
+       ' for every grouping if its deduction changed' do
+      deductive_text.update!(deduction: 2.0)
+      assignment.reload
+      marks = []
+      assignment.groupings.includes(:current_result).each do |grouping|
+        marks << grouping.current_result.marks
+                         .find_by(markable_id: annotation_category_with_criterion.flexible_criterion_id).mark
```

Comment:
Since the markable association is polymorphic, it's better to do `find_by(markable: annotation_category_with_criterion.flexible_criterion)` (avoids any possibility of duplicate ids).